### PR TITLE
SVD docs improved

### PR DESCRIPTION
**Pull request by lezcano**
- Corrected a few errata in the SVD docs
- Made the notation more uniform (refer to `Vh` in `linalg.svd`, always use double tilts...)
- Wrote a better explanation about why the gradients of `U` and `V` are not well-defined when the input is complex or real but has repeated singular values. The previous one pointed to a somewhat obscure post on gauge theory.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Maybe worth preserving the link to gauge theory blog in some Notes section?

**Comment by Lezcano** [comment]
I explained the problems in simpler terms. In my opinion, the post was too abstract for a person not used to the notation in differential geometry / physics.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8471]: "Fortran"

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8465]: "SVD" -> ":func:\`torch.svd\`"

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8466]: "SVD" -> ":func:\`torch.svd\`"

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8483]: "SVD" -> "singular value decomposition"

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8488]: Numpy -> NumPy

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8490]: For torch.linalg.svd we should clarify what actually happens. When say the gradient is only "well-defined" in certain cases we're not actually telling the reader what torch.svd() or torch.linalg.svd() does. Does it compute the gradient when the gradient is well-defined? What does it compute when the gradient is not well-defined? Does it throw a runtime error?

Comment (mruberry) [inline torch/_torch_docs.py:8490]: "the backward operation" -> "the gradient"?

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:685]: Analogous suggestions here as above

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:690]:
```
.. note:: Each \`\`U\`\` matrix will be transposed (equivalently "column-major" or "Fortran-contiguous").
```

U may not be just a matrix, and it's odd that this note started by explaining what U was not instead of what it was. Saying the matrix or matrices are transposed may be more familiar to PyTorch users.

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:693]: What does "unstable" mean?

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:695]: Not a list.

"See also the below note on complex gradients"?

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:723]: torch.Tensor() is deprecated. How about torch.tensor()?

**Comment by mruberry** [comment]
I like a lot of what this PR does and think it does make the writing clearer. The mathematical formatting is very jarring, however. Check out: https://11545664-65600975-gh.circle-artifacts.com/0/docs/linalg.html. Maybe we can make the inline mathematical expressions more natural with the surrounding text?

There are a few places where I would like our docs to be clearer, but this PR doesn't need to improve upon them.

**Comment by IvanYashchuk** [comment]
Links to rendered documentation from this PR: 
[torch.linalg.svd](https://11545664-65600975-gh.circle-artifacts.com/0/docs/linalg.html#torch.linalg.svd)
[torch.svd](https://11545664-65600975-gh.circle-artifacts.com/0/docs/generated/torch.svd.html?highlight=torch%20svd#torch.svd)

Screenshots of the main text
Before:
image
After:
image
I have concerns about the rendering of equations and code highlighting.

File: torch/linalg/__init__.py
Comment (Lezcano) [reply torch/linalg/__init__.py:690]: The problem with this is that if you say that "U is transposed", some people might do `U, _, V = torch.svd(X); U = U.T` to "account for that".

I think that the people that will care about the strides of `U` will already know what strides are and what a contiguous tensor is. Everyone else will just omit this note.

About `U` not being a matrix, fair enough. I just used the word "matrix" because the previous note used it. I can change that to "The returned `U` will not be contiguous. The matrix (or batch of matrices) will be represented as column-major matrices (i.e. Fortran-contiguous)." What do you think?

File: torch/linalg/__init__.py
Comment (Lezcano) [reply torch/linalg/__init__.py:693]: Close to singular. Basically the gradient has a `1. / (s_i - s_j)` term, where `s_i` are the singular values.

**Comment by Lezcano** [comment]
About the code rendering, fair enough. I will change it back to double back quotes and unicode.

About the code highlighting, @IvanYashchuk raised the point that when you write equations such as `(m, n)`, the code generated (not in this comment but in the docs) has an odd white space in the middle. I believe that this is not a problem coming from PyTorch but from sphinx, so there is not much to do about that but hope that the rendering from sphinx gets better in the future.

I think it is better to unify everything that's a variable / formula to have a double back tilt. I believe this is important as, for example, in this SVD documentation, we sometimes write `input` to refer to the argument, and sometimes we use it as a noun. This clearly deambiguates both usages. I also prefer it over the single tilt, as the single tilt sometimes goes unnoticed, and it also not consistent with the highligh given by `:attr:`. What do you think @mruberry @IvanYashchuk ?

**Comment by IvanYashchuk** [comment]
My suggestions for torch.linalg documentation formatting:

- Use of TeX:
  - use Unicode instead of inline TeX math for consistent font and font-size
  - TeX math mode can be used for separate blocks of complex mathematical expressions (for example [torch.kron](https://pytorch.org/docs/master/generated/torch.kron.html)), not for the inline math text
- Use of code highlighting (\`\`code_here\`\`):
  - for inline text use it only if it's really necessary
  - use it only for real code (for example [torch.linalg.tensorsolve](https://pytorch.org/docs/stable/linalg.html?highlight=tensorsolve#torch.linalg.tensorsolve)) preferably without spaces (or fix breaking of code highlighting in sphinx settings somewhere)
  - for all function arguments use :attr:\`arg\`
  - use _True_, _False_, _None_ instead of `True`, `False`, `None`
  - use _A_ for mathematical variables instead of `A`

I don't `think` this kind `of` inline highlighting `looks` good `in` the text and helps `drawing` attention. 
- Use of technical references:
  - If the resource is not widely known (not a Wikipedia article), then it's better to have a link to resources, that better and in more details explain certain concepts and facts of linear algebra (in this case I'd keep the link about the gauge problem in automatic differentiation)

**Comment by Lezcano** [comment]
I agree about the "Use of TeX" part.

**About the code highlighting,** in my opinion, it's better to always go for double back tilts.
Pros:
- It's uniform. If we use single \`A\` for variables, :attr:\`arg\` for arguments then we have a mix of the two. This gets even worse if we want to write a formula involving \`A\` and :attr:\`arg\`, where half of the formula has grey background and the other half is in italics. I argue that this mix of styles is much more distracting than evening them all out to using the grey background.

Cons:
- It has spaces. I think this should be solved by configuring Sphinx.

**About the technical references**, I agree. That being said, I do not think that the link to those formulas on gauges are very useful here. I think that the amount set of people that do not know that the SVD is not unique but do know what a gauge is and would not be scared by the following formula is basically empty

image

I believe that this link should go as a comment in `svd_backward` as a reference of the implementation followed.

Surely you can explain this lack of uniqueness as a gauge-invariance problem, as the Stiefel manifold is a principal bundle over the Grassmannian, but this is overkill. I think is better to explain that what's happening is that the SVD chooses subspaces, not bases of orthonormal vectors, and to represent those subspaces we need to choose a basis, but we can choose any basis. In particular we can choose a basis and rotate it inside this subspace and that will give another basis of the subspace. This is even explained in the wikipedia article of the SVD

https://en.wikipedia.org/wiki/Singular_value_decomposition#Singular_values,_singular_vectors,_and_their_relation_to_the_SVD

> In general, the SVD is unique up to arbitrary unitary transformations applied uniformly to the column vectors of both U and V spanning the subspaces of each singular value, and up to arbitrary unitary transformations on vectors of U and V spanning the kernel and cokernel, respectively, of M.

This all boils down to choosing a level of mathematical background that we are assuming on the reader and write catering to that level. When it comes to linear algebra, I think it is safer to stay on the conservative side.

**Comment by mruberry** [comment]
> I agree about the "Use of TeX" part.
> 
> **About the code highlighting,** in my opinion, it's better to always go for double back tilts.
> Pros:
> 
> * It's uniform. If we use single `A` for variables, :attr:`arg` for arguments then we have a mix of the two. This gets even worse if we want to write a formula involving `A` and :attr:`arg`, where half of the formula has grey background and the other half is in italics. I argue that this mix of styles is much more distracting than evening them all out to using the grey background.
> 
> Cons:
> 
> * It has spaces. I think this should be solved by configuring Sphinx. For example, writing things like `(m, n)` on github does not produce any awkward space, so there's some hope there.

I welcome the debate on how best to format the docs but would like to suggest we move it to its own issue and not block this PR. That leaves the question, then, of how do we format this PR. Would it be OK, @Lezcano, if we try to minimize the disruption of weird whitespace for now, follow-up fixing that issue, create a standard, and then possibly revisit this to impose that standard? 

> **About the technical references**, I agree. That being said, I do not think that the link to those formulas on gauges are very useful here. I think that the amount set of people that do not know that the SVD is not unique but do know what a gauge is and would not be scared by the following formula is basically empty
> 
> image
> 
> I believe that this link should go as a comment in `svd_backward` as a reference of the implementation followed.
> 
> Surely you can explain this lack of uniqueness as a gauge-invariance problem, as the Stiefel manifold is a principal bundle over the Grassmannian and the SVD decomposition is a map onto a product of Grassmannians, but this is overkill. I think is better to explain that what's happening is that the SVD chooses subspaces, not bases of orthonormal vectors, and to represent those subspaces we need to choose a basis, but we can choose any basis. In particular we can choose a basis and rotate it inside this subspace and that will give another basis of the subspace. This is even explained in the wikipedia article of the SVD
> 
> https://en.wikipedia.org/wiki/Singular_value_decomposition#Singular_values,_singular_vectors,_and_their_relation_to_the_SVD
> 
> > In general, the SVD is unique up to arbitrary unitary transformations applied uniformly to the column vectors of both U and V spanning the subspaces of each singular value, and up to arbitrary unitary transformations on vectors of U and V spanning the kernel and cokernel, respectively, of M.
> 
> This all boils down to choosing a level of mathematical background that we are assuming on the reader and write catering to that level. When it comes to linear algebra, I think it is safer to stay on the conservative side.

I agree with you on the complexity of the "Gauge-based" explanation vs. one using bases, which are familiar to anyone who's taken a course in linear algebra.

**Comment by Lezcano** [comment]
Sure, let's move this discussion somewhere else.

That being said, I do not know when to use the back tilts. I see these cases:
- If the variable is an attribute: use :attr:\`var\` 
- If the variable is not an attribute, such as return variables, or it's a constant like True or False: use single back tilts.
- If we have a maths formula: use single back tilts
- If we have some short code:
  -  If it just involves non-input variables: use double tilts
  -  If it also involves input variables (such as `compute_uv = True`): what do we do here?

For the last case, should it be:
- :attr:\`compute_uv\` \`= True\`  // This would use two different fonts one after the other
- :attr:\`compute_uv\` = \`True\`   // This would use three different fonts one after the other. This was the way it was done before.
- :attr:\`compute_uv\` \`\`=True\`\` // This would use just one font, but the font of `True` would be different to the usual font given by rule 2. above.

Again, I'm happy to format this PR in whichever way you see fit. I just do not know what should be the formatting.

File: torch/_torch_docs.py
Comment (Lezcano) [reply torch/_torch_docs.py]: That is a great question. The answer is: I do not know.

I have been trying to undestand this problem better, which lead to the discussion in https://github.com/pytorch/pytorch/issues/47599#issuecomment-800920251
It happens to be a surprisingly tricky problem, and I do not understand it well-enough just yet. I will come back to this once I do.

Comment (mruberry) [reply torch/_torch_docs.py]: Appreciating what we don't know is a great start.

Maybe we can unblock this PR by:

- leaving this issue open for future resolution
- saying that produced gradients are undefined unless the grad passed to svd has certain properties (this would require some wordsmithing)

File: torch/linalg/__init__.py
Comment (mruberry) [reply torch/linalg/__init__.py:689]: I appreciate your thinking and your suggestion sounds good.

File: torch/linalg/__init__.py
Comment (mruberry) [reply torch/linalg/__init__.py:692]: Can we clarify that in the documentation? Or maybe we should consider adding a glossary to the linalg docs?

**Comment by mruberry** [comment]
> Sure, let's move this discussion somewhere else.
> 
> That being said, I do not know when to use the back tilts. I see these cases:
> 
> * If the variable is an attribute: use :attr:`var`
> * If the variable is not an attribute, such as return variables, or it's a constant like True or False: use single back tilts.
> * If we have a maths formula: use single back tilts
> * If we have some short code:
>   
>   * If it just involves non-input variables: use double tilts
>   * If it also involves input variables (such as `compute_uv = True`): what do we do here?
> 
> For the last case, should it be:
> 
> * :attr:`compute_uv` `= True`  // This would use two different fonts one after the other
> * :attr:`compute_uv` = `True`   // This would use three different fonts one after the other. This was the way it was done before.
> * :attr:`compute_uv` ``=True`` // This would use just one font, but the font of `True` would be different to the usual font given by rule 2. above.
> 
> Again, I'm happy to format this PR in whichever way you see fit. I just do not know what the formatting should be.
> Edit. cc @mruberry @IvanYashchuk

We really should come up with a style guide... I actually prefer minimizing the use of backticks, but in this case I would maximize uniformity by writing one mathematical expression (ditching the :attr:) or doing :attr\`compute_uv\`=True. We don't need to do things perfectly in this PR; we'll have a chance to review the docs holistically ahead of 1.9.

File: torch/linalg/__init__.py
Comment (Lezcano) [reply torch/linalg/__init__.py:692]: I've added a better explanation now as its done in the documentation of `symeig`. I will open an issue to discuss how to make the documentation of `torch.linalg` more consistent, to discuss both the style and things like this one.

**Comment by Lezcano** [comment]
I have done the following changes:
- Prefer single backtiles over double backtiles.
- Simpler explanation of the problem with the gradients, leaving space to be modified in the future if we come up with a better formula
- Ordered the notes / warnings from simpler to more specific
@mruberry @IvanYashchuk could you take a look at the current state and tell me what you think?

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8472]: I didn't follow this part as closely as the rest of the documentation, but I think it's an improvement and we should stop this PR here and revisit final touches to the documentation later.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8485]: We can refine this language, but like the above I think that's a problem for the future.

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:8494]: This is actually the only blocking issue I have with the PR: the "#noqa: E501 tag" is printing in the actual docs.

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:880]: Analogous comments to the torch.svd() comments (as expected), but again the blocking issue here is the printing of this comment.

**Review by mruberry — COMMENTED** [review]
This looks like a significant improvement and I'd like to merge it. The #noqa is printing, however.

@IvanYashchuk would you like to take a look again? Not everything has to be perfect at this time. We should plan a doc review before the PyTorch 1.9 branch cut.

File: torch/_torch_docs.py
Comment (Lezcano) [reply torch/_torch_docs.py:8472]: Note that this was already in the previous documentation. I just reordered the points from simpler to more specific.

File: torch/_torch_docs.py
Comment (Lezcano) [reply torch/_torch_docs.py:8485]: I completely agree. I took and adapted this part from a note in `symeig`. I did not change the wording to keep things simpler. I will be happy to improve a bit the wording later down the road, when I go over all the `torch.linalg` operations.

File: torch/_torch_docs.py
Comment (IvanYashchuk) [inline torch/_torch_docs.py:8494]: Could you move this link to the end similarly to how it was done for `Gauge problem in AD` link.
See https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages

**Review by IvanYashchuk — APPROVED** [review]
I think it's good to be merged. Thank you, Mario, for doing this!

**Comment by Lezcano** [comment]
Sorry for that `#nolint` issue. It should be good now.
I'll create in a second an issue to discuss the docs of `torch.linalg` in more generality. @mruberry

**Comment by Lezcano** [comment]
I have not been able to split the long line. I really hope that flake8 does not complain. If it does, I do not know how to solve it. I have found [this SO post](https://stackoverflow.com/questions/14414878/how-do-i-break-a-link-in-a-rst-docstring-to-satisfy-pep8), but I have not managed to replicate the behaviour that they show there.

**Comment by Lezcano** [comment]
Moving the discussion about the formatting here: https://github.com/pytorch/pytorch/issues/54878

**Review by mruberry — APPROVED** [review]
Cool! Always great to have better docs.

**Comment by Lezcano** [comment]
Wow, that's a new one. I'll have a look in a second.

**Pull request closed without merging**
